bulk_plate_motion.py

yuankailiu opened the pull request:
+ add a code for bulk plate motion estimation/correction

+ reference_point.py: use FILE_TYPE for velocity datasetName

**Description of proposed changes**

## Major change
Add a code for the estimation of ground bulk motion. It is meant to account for the apparent bulk-motion ramp in the velocity field due to the reference frame of the satellite. The bulk motion can be estimated by either:

1. assuming a constant translation (manually input ve, vn, vu based on your prior knowledge)
2. using a plate rotation model (manually input Euler Pole rotation)

The plate rotation calculation from Euler Pole to ENU velocity on the grid is based on this python package: https://github.com/lcx366/PlateTectonic

Please see the usage
`bulk_plate_motion -h`

Also need to install the dependencies of that `platemotion` package by:
1. `pip install platemotion`
2. Figure out how to do it by cloning github and install via conda

## Minor change

+ reference_point.py: use FILE_TYPE for velocity datasetName

**Reminders**

- [ ] Fix #xxxx
- [ ] Pass Codacy code review (green)
- [ ] Pass Circle CI test (green)
- [ ] Make sure that your code follows our style. Use the other functions/files as a basis.
- [ ] If modifying functionality, describe changes to function behavior and arguments in a comment below the function declaration.
- [ ] If adding new functionality, add a detailed description to the documentation and/or an example.

yunjunz commented:
Thank you @yuankailiu for this very cool PR!

For the output plate motion file, let's put them into 2 files:
+ `inputs/BulkPlateMotion3D.h5` for the calculated plate motion in east/north/up direction.
+ `inputs/BulkPlateMotion.h5` for the LOS bulk motion file.

Both of them are spatially absolute. Then one could run `diff.py velocity.h5 inputs/BulkPlateMotion.h5 -o velocity_BPM.h5` to get the corrected velocity while applying the spatial referencing on the fly, without physically writing a spatially reference BulkPlateMotion.h5 file.

I will go through the rest of the PR at another time.

yuankailiu commented:
## Updates

### Addressed PR comments
 
(https://github.com/insarlab/MintPy/pull/769#issuecomment-1131231755)

For the output plate motion file, save them into 2 files:
+ inputs/BulkPlateMotion3D.h5: for the calculated plate motion in east/north/up direction.
+ inputs/BulkPlateMotion.h5: for the LOS bulk motion file.
Both of them are spatially absolute.

The diff.py operation will easily leverage the "spatial referencing on the fly" when apply to an input velocity file.

For the corrected velocity file, use a universal suffix of _BPM, e.g. velocity_BPM.h5


### diff.py: reference point (mintpy/diff.py)
Check and apply the reference point for the velocity files

yunjunz commented:
It seems that part of the changes in `reference_point.py` in this PR is duplicated in #780, which is now merged. @yuankailiu Could you rebase this PR against the latest main branch?

yunjunz commented on mintpy/bulk_plate_motion.py:
```diff
@@ -0,0 +1,369 @@
+  bulk_plate_motion.py -g inputs/geometryGeo.h5 -v velocity.h5 --om_sph  54.45 259.66 0.255   -m waterMask.h5
+  bulk_plate_motion.py -g inputs/geometryGeo.h5 -v velocity.h5 --om_cart 1.154 -0.136 1.444   -m maskTempCoh.h5
+  bulk_plate_motion.py -g inputs/geometryGeo.h5 -v velocity.h5 --enu     25.0 30.5 0.0        -m zero
+"""
```
Two questions here for usage:

1. Should we use ONE AND ONLY ONE of `--om_sph`, `--om_cart` and `--enu` to specify the plate motion?

2. For the detailed values you used for each of the 3 options. Could you please a) add a comment above each command line, to specify for which plate these example values are? b) write a note/procedure on how to get these values, if one want to apply this correction to their own dataset?

yuankailiu replied:
1. Yes, we use one and only one. I will further write that in the usage
2. I will do it now, sounds good!

yuankailiu commented:
> It seems that part of the changes in `reference_point.py` in this PR is duplicated in #780, which is now merged. @yuankailiu Could you rebase this PR against the latest main branch?

Yes, I will do it now.

yunjunz commented on mintpy/bulk_plate_motion.py:
```diff
@@ -0,0 +1,372 @@
+  Peter, H., Fernández, M., Aguilar, J., & Fernández, J. (2021). Copernicus POD Product Handbook: 
+    Copernicus Sentinel-1, -2 and -3 Precise orbit Determination Serivice (CPOD) (GMV-CPOD-TN-0009). 
+    https://sentinels.copernicus.eu/documents/247904/3372484/Sentinels-POD-Product-Handbook-1.19.pdf
+"""
```
Please add 1) a brief description and 2) a complete reference for each of the 3 plate motion models (NNR-NUVEL1A, NNR-MORVEL56, NNR-ITRF14) here, as we discussed offline, to help users to decide which model they may want to use.

yunjunz reviewed (approved):
I have finished my part of the refactoring. The PR looks good to me, pending the minor comment above.

Thank you @yuankailiu for this very cool and useful PR!
